tools/unittestharness.py

```diff
@@ -1,120 +1,205 @@
+                    failMsg += 'Empty message'
+                if testOutput[:4] == 'Warn':
+                    xmlEntry.add_failure_info(message=failMsg.lstrip(),
+                                              failure_type='warning')
+                else:
+                    xmlEntry.add_failure_info(message=failMsg.lstrip())
+            except AttributeError:
+                pass
+
+            if testProc.stderr:
+                xmlEntry.stderr = testProc.stderr
+
+            if otherOut:
+                xmlEntry.stdout = otherOut[:-1]
+                otherOut = ''
+
+            xmlParser.test_cases.append(xmlEntry)
+
+    printResults(testsResults, errorList, skipList)
+
+    with open(xml_outfile, 'w') as fid:
+        xmlParser.to_file(fid, [xmlParser])
+
+
+def updateEnvVar(envVar, envPath):
     try:
-        os.environ["LD_LIBRARY_PATH"] = os.getcwd() + os.sep + sys.argv[1] + os.sep + "lib:" + os.environ["LD_LIBRARY_PATH"]
+        if str(envPath) not in os.environ[envVar]:
+            os.environ[envVar] += f':{envPath}'
     except KeyError:
-        os.putenv("LD_LIBRARY_PATH", os.getcwd() + os.sep + sys.argv[1] + os.sep + "lib")
-
-    if "--electricfence" in sys.argv:
-      os.putenv("LD_PRELOAD", "/usr/lib/libefence.so.0.0")
-      #os.putenv("EF_DISABLE_BANNER", "1")
-
-    TestHarness = UnitTestHarness(sys.argv[-1])
-    TestHarness.run()
+        os.environ[envVar] = str(envPath)
+
+
+parser = argparse.ArgumentParser(description='''Fluidity UnitTestHarness''')
+parser.add_argument('-d', '--dir', default='bin/tests', metavar='',
+                    help="""directory where to look for UnitTest binary(ies) -
+                            defaults to '%(default)s'""")
+parser.add_argument('-t', '--test', metavar='',
+                    help='''single UnitTest to run - should match an existing
+                            UnitTest within --dir''')
+parser.add_argument('-f', '--from-file', metavar='',
+                    help='''file containing UnitTests to run - expects one test
+                            entry per line''')
+parser.add_argument('-x', '--xml-output', metavar='',
+                    help='''filename where to write the xml output - junit_xml
+                            is required to generate the xml output''')
+parser.add_argument('-c', '--clean', action='store_true',
+                    help='remove the directory provided through --dir')
+parser.add_argument('--efence', action='store_true',
+                    help='links against libefence')
+args = parser.parse_args()
+
+pyPath = (Path(sys.argv[0]).parent / '../python').resolve()
+updateEnvVar('PYTHONPATH', pyPath)
+updateEnvVar('LD_LIBRARY_PATH', (pyPath / '../lib').resolve())
+if args.efence:
+    updateEnvVar('LD_PRELOAD', Path('/usr/lib/libefence.so.0.0'))
+# os.environ['EF_DISABLE_BANNER'] = '1'
+
+testsDir = pyPath / f'../{args.dir}'
+assert testsDir.is_dir(), f'{testsDir.resolve()} is not a valid directory'
```

Comment:
Makes more sense as well. However, I have added a check in case we are not running from fluidity's root directory and args.dir is the default. Otherwise that would force the user to be in fluidity's root directory to execute unittestharness or to provide a path to the default directory, which I believe is slightly annoying.